lib/src/core/security_provider.dart

```diff
@@ -26,7 +29,22 @@ import 'credentials/psk_credentials.dart';
 typedef ClientPskCallback = PskCredentials? Function(
   Uri uri,
   Form? form,
-  String? identityHint,
+  Uint8List? identityHint,
+);
+
+/// Function signature for a synchronous callback for providing client
+/// [AceCredentials] at runtime, based on an optional [creationHint]
+/// given by the Resource Server. This creation hint has to be parsed by the
+/// library user.
+///
+/// If a request with an access token has failed before, leading to an
+/// "Unauthorized" response, the [invalidAceCredentials] from the previous
+/// request are returned as an additional parameter.
+typedef AceSecurityCallback = Future<AceCredentials?> Function(
+  Uri uri,
+  Form? form,
+  List<int>? creationHint,
+  AceCredentials? invalidAceCredentials,
 );
```

Comment:
In order to achieve a "dependency agnostic" solution while considering the complexity of the `AuthServerRequestCreationHint` class, I now tend towards passing the creation hint in its serialized `List<int>` form to the user. Do you think this would be viable? Or should the `AuthServerRequestCreationHint` class be re-implemented instead?

Reply:
By "dependency agnostic", do you mean not depending on `dart_dcaf` in this case? If so, I don't quite see the reason for it to be honest. It seems more fitting to me to "outsource" the ACE-OAuth data models to `dart_dcaf`, so that we can simply use the `AuthServerRequestCreationHint` class here. I also highly doubt we'll see a second ACE-OAuth implementation in Dart, but if a user does want to use a hypothetical library other than `dart_dcaf`, they could still call `.serialize()` on the data model provided by `dart_dcaf` and then pass it to the other library.

Perhaps I'm missing some reasoning behind it, so if you do decide on not using `dart_dcaf` for it, I'd strongly recommend the serialized approach, otherwise we'd have to basically duplicate code from `dart_dcaf` (i.e. data models, serialization, equality, all the `Scope` derivatives and so on) to reimplement the creation hint class.

Reply:
Hmm, yeah, you are right, maybe I am overthinking this a little. I'll experiment a bit more with the serialized approach in fixup commits but if it doesn't make sense then we can simply drop the additional commits (including the latest one already present on this branch).

Reply:
Thank you for the feedback already, in any case :)